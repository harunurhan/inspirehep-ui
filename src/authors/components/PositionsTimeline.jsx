import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';
import { Timeline } from 'antd';
import { hasAnyOfKeys } from '../../common/utils';

class PositionsTimeline extends Component {
  static renderPositionTimelineItem(position) {
    const institution = position.get('institution');
    const rank = position.get('rank');
    const dateDisplay = PositionsTimeline.getPositionDateDisplay(position);

    return (
      <Timeline.Item>
        <div>{dateDisplay}</div>
        <div>
          {rank && <strong>{rank}, </strong>}
          <span>{institution}</span>
        </div>
      </Timeline.Item>
    );
  }

  static getPositionDateDisplay(position) {
    const current = position.get('current');
    const startDate = position.get('start_date');
    const endDate = position.get('end_date');

    const suffixedStartDateOrEmpty = startDate ? `${startDate}-` : '';

    if (current) {
      return `${suffixedStartDateOrEmpty}present`;
    }

    if (endDate) {
      return `${suffixedStartDateOrEmpty}${endDate}`;
    }

    if (startDate) {
      return startDate;
    }

    return null;
  }

  shouldDisplayTimeline() {
    const { positions } = this.props;
    if (positions.size > 1) {
      return true;
    }

    const position = positions.first();
    return (
      !position.get('current') ||
      hasAnyOfKeys(position, ['start_date', 'end_date', 'rank'])
    );
  }

  render() {
    const { positions } = this.props;
    return (
      this.shouldDisplayTimeline() && (
        <Timeline>
          {positions.map(PositionsTimeline.renderPositionTimelineItem)}
        </Timeline>
      )
    );
  }
}

PositionsTimeline.propTypes = {
  positions: PropTypes.instanceOf(List),
};

PositionsTimeline.defaultProps = {
  positions: null,
};

export default PositionsTimeline;
