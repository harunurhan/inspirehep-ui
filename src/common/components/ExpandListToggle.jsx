import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SecondaryButton from './SecondaryButton';

class ExpandListToggle extends Component {
  render() {
    const { size, limit, onToggle, expanded } = this.props;

    if (size <= limit) {
      return null;
    }

    const buttonText = expanded ? 'Hide' : `Show all (${size})`;
    return <SecondaryButton onClick={onToggle}>{buttonText}</SecondaryButton>;
  }
}

ExpandListToggle.propTypes = {
  size: PropTypes.number.isRequired,
  limit: PropTypes.number.isRequired,
  onToggle: PropTypes.func.isRequired,
  expanded: PropTypes.bool.isRequired,
};

export default ExpandListToggle;
