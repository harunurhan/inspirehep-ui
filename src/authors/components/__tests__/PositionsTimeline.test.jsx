import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import PositionsTimeline from '../PositionsTimeline';
import ExpandListToggle from '../../../common/components/ExpandListToggle';

describe('PositionsTimeline', () => {
  it('renders with positions', () => {
    const positions = fromJS([
      {
        institution: 'Inst 1',
        display_date: '1990-1994',
        rank: 'UNDERGRADUATE',
      },
      {
        institution: 'Inst 2',
        display_date: '1994-2000',
        rank: 'PHD',
      },
      {
        institution: 'CERN',
        display_date: '2000-present',
        rank: 'STAFF',
        current: true,
      },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with single position if has rank even if current', () => {
    const positions = fromJS([
      {
        institution: 'Inst 1',
        rank: 'UNDERGRADUATE',
        display_date: 'present',
        current: true,
      },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with single position if has display_date even if current', () => {
    const positions = fromJS([
      {
        institution: 'Inst 1',
        display_date: '2005',
        current: true,
      },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render with single position without any of extra display keys if current', () => {
    const positions = fromJS([
      {
        institution: 'Inst 1',
        current: true,
      },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without any of extra display keys if not current', () => {
    const positions = fromJS([
      {
        institution: 'Inst 1',
        current: false,
      },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders only first 5 by default', () => {
    const positions = fromJS([
      { institution: 'Inst 1' },
      { institution: 'Inst 2' },
      { institution: 'Inst 3' },
      { institution: 'Inst 4' },
      { institution: 'Inst 5' },
      { institution: 'Inst 6' },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders all when ExpandListToggle is toggled by default', () => {
    const positions = fromJS([
      { institution: 'Inst 1' },
      { institution: 'Inst 2' },
      { institution: 'Inst 3' },
      { institution: 'Inst 4' },
      { institution: 'Inst 5' },
      { institution: 'Inst 6' },
    ]);
    const wrapper = shallow(<PositionsTimeline positions={positions} />);
    const toggleWrapper = wrapper.find(ExpandListToggle);
    const onExpandToggle = toggleWrapper.prop('onToggle');
    onExpandToggle();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
