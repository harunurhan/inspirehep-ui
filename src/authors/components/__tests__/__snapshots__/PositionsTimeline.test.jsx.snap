// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionsTimeline does not render with single position without any of extra display keys if current 1`] = `""`;

exports[`PositionsTimeline renders with positions 1`] = `
<Timeline
  mode=""
  prefixCls="ant-timeline"
  reverse={false}
>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div>
      1990-1994
    </div>
    <div>
      <strong>
        UNDERGRADUATE
        , 
      </strong>
      <span>
        Inst 1
      </span>
    </div>
  </TimelineItem>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div>
      1994-2000
    </div>
    <div>
      <strong>
        PHD
        , 
      </strong>
      <span>
        Inst 2
      </span>
    </div>
  </TimelineItem>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div>
      2000-present
    </div>
    <div>
      <strong>
        STAFF
        , 
      </strong>
      <span>
        CERN
      </span>
    </div>
  </TimelineItem>
</Timeline>
`;

exports[`PositionsTimeline renders with single position if has end_date even if current 1`] = `
<Timeline
  mode=""
  prefixCls="ant-timeline"
  reverse={false}
>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div>
      present
    </div>
    <div>
      <span>
        Inst 1
      </span>
    </div>
  </TimelineItem>
</Timeline>
`;

exports[`PositionsTimeline renders with single position if has rank even if current 1`] = `
<Timeline
  mode=""
  prefixCls="ant-timeline"
  reverse={false}
>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div>
      present
    </div>
    <div>
      <strong>
        UNDERGRADUATE
        , 
      </strong>
      <span>
        Inst 1
      </span>
    </div>
  </TimelineItem>
</Timeline>
`;

exports[`PositionsTimeline renders with single position if has start_date even if current 1`] = `
<Timeline
  mode=""
  prefixCls="ant-timeline"
  reverse={false}
>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div>
      2000-present
    </div>
    <div>
      <span>
        Inst 1
      </span>
    </div>
  </TimelineItem>
</Timeline>
`;

exports[`PositionsTimeline renders without any of extra display keys if not current 1`] = `
<Timeline
  mode=""
  prefixCls="ant-timeline"
  reverse={false}
>
  <TimelineItem
    color="blue"
    pending={false}
    prefixCls="ant-timeline"
  >
    <div />
    <div>
      <span>
        Inst 1
      </span>
    </div>
  </TimelineItem>
</Timeline>
`;
